Extend ROS resource addressing design document

hidmic opened the pull request:
This pull request generalizes the existing topic and service name addressing design document to cover the current goals in terms of addressing topics, services, actions and parameters in ROS 2.

hidmic committed:
Refactor and extend ROS resource addressing design document.

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

hidmic commented:
As a follow-up, the design document on static name remapping will need an update as well.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -1,10 +1,10 @@
-title: Topic and Service name mapping to DDS
-permalink: articles/topic_and_service_names.html
+title: Resource Names
+permalink: articles/resource_names.html
```
I'm hesitant to rename the permalink, maybe we can setup a redirect? I just don't want to break existing links if we can help it.

hidmic replied:
Sure, we can! See redirect in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -19,13 +19,33 @@ published: true
+### Concepts
 
-### ROS 1 Names
+#### Content-oriented resources
```
I think it would be good to motivate our use of this terminology (and potentially change it if something else fits better). There is some stuff on wikipedia, but it's mostly focused on specific technologies like actually implementing internet networks using named resources or are tightly coupled to web technologies:

- https://en.wikipedia.org/wiki/Content-addressable_storage#Content-addressed_vs._location-addressed
- https://en.wikipedia.org/wiki/Content_centric_networking
- https://en.wikipedia.org/wiki/Information-centric_networking
- https://en.wikipedia.org/wiki/Named_data_networking
- https://en.wikipedia.org/wiki/Resource-oriented_architecture

I had trouble finding an article which summarized the basic difference in addressing between having a host and not. The URI page talks about host and path, and that host is optional:

https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Definition

I don't think we need to spend much time here, but a few citations might lend credibility to the distinction we're making here.

hidmic replied:
> I think it would be good to motivate our use of this terminology

I agree, and I think that CCN (Content Centric Networking) gets pretty close. A related [Linux Foundation project](https://wiki.fd.io/view/Cicn) adds Publish/Subscribe Networking as a potential alias for the same concept. See 7680627ee8f2fa8675fd58e72f814585c6721f2a.

> The URI page talks about host and path, and that host is optional:

For URIs, a host is always defined, either explicitly or implicitly through a sane default e.g. `localhost` for `file://` URIs, see [RFC 3986 Section 3.2.2](https://tools.ietf.org/html/rfc3986#section-3.2.2).

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -19,13 +19,33 @@ published: true
+#### Host-oriented resources
+
+Host-oriented resources are provided by a node in a given system and thus are bound to it.
+Multiple resources with the same name and type may by provided by different nodes.
```
```suggestion
Multiple resources with the same name and type may be provided by different nodes.
```

hidmic replied:
Corrected in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -19,13 +19,33 @@ published: true
+Therefore, the provider node name is necessary to unambiguously address a host-oriented resource.
+In the preceding, it is assumed that node name uniqueness has been ensured by means not prescribed in this document.
+
+ROS services, actions and parameters fall under this classification.
```
```suggestion
ROS services, actions, and parameters fall under this classification.
```

Subjective, but we use the Oxford comma in other documents, so I recommend it here for consistency.

hidmic replied:
Sounds good. Added in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -38,152 +58,157 @@ A valid name has the following characteristics:
-- must not contain any number of repeated underscores (`_`)
-- must separate a tilde (`~`) from the rest of the name with a forward slash (`/`), i.e. `~/foo` not `~foo`
-- must have balanced curly braces (`{}`) when used, i.e. `{sub}/foo` but not `{sub/foo` nor `/foo}`
+#### Provider Names
```
This goes along with the other comment I made, I think we should consider adopting the URI terminology where it makes sense. Would `host` make more sense than `provider` in this case? I'm not 100% sure, but I'd like to know your opinion.

hidmic replied:
So, the rationale here is that since nodes provide services instead of hosting data it made more sense to use the former terminology, but the distinction is subtle. I'm fine with `host` too. See 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood replied:
Yeah, I don't feel strongly about it, host is fine and I would be ok with provider as well so long as there is a rationale and it's not arbitrary.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -38,152 +58,157 @@ A valid name has the following characteristics:
 
-The content of substitutions, i.e. the string in side of balanced curly braces (`{}`), follow very similar rules to names.
-The content of substitutions:
+For host-oriented resources i.e. services, actions and parameters, the provider name refers to the node that is exposing said resources.
```
```suggestion
For host-oriented resources, i.e. services, actions, and parameters, the provider name refers to the node that is exposing said resources.
```

hidmic replied:
Corrected in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -19,13 +19,33 @@ published: true
+Content-oriented resources are not bound to any particular node in a given system.
+ROS topics, which may be freely published or subscribed to by multiples ROS nodes, fall under this classification.
+
+Note that in ROS 1, every resource i.e. topics, services, actions and parameters, is a content-oriented resource.
```
```suggestion
Note that in ROS 1, every resource, i.e. topics, services, actions, and parameters, is a content-oriented resource.
```

hidmic replied:
Corrected in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -38,152 +58,157 @@ A valid name has the following characteristics:
+##### Definition
 
-### Fully Qualified Names
+A provider name is made up of one or more tokens, delimited by periods (`.`) i.e. using URL hostname style syntax.
```
The structure of this is covered by FQDN's I think (https://en.wikipedia.org/wiki/Fully_qualified_domain_name) if you want another citation.

wjwwood replied:
We use the "fully qualified [node] name" phrase in places already.

hidmic replied:
Corrected in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -38,152 +58,157 @@ A valid name has the following characteristics:
-- must not contain tilde (`~`) or curly braces (`{}`)
+- must not be empty.
+- may use alphanumeric characters (`[0-9|a-z|A-Z]`) and/or underscores (`_`) only.
+- must not start with numeric characters (`[0-9]`).
```
Are they also not allowed to have `__` in the name? I think that might be a restriction for node names and/or namespaces right now.

hidmic replied:
That's right. Added missing constraint in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -38,152 +58,157 @@ A valid name has the following characteristics:
+#### Resource Names
 
-For example, these are valid names:
+All resources, whether content-oriented or host-oriented, have a name.
```
I like name myself, but again looking to converge on "standard" terminology, I believe this would be referred to as the `path` in URI parlance. Perhaps just an aside the first time we mention it (here)? Like `have a name (the 'path' of the URI)`? I don't know what would be more appropriate, I just wanted to bring it up.

hidmic replied:
I agree with establishing a relationship between resource name and path, but I wouldn't switch to URI parlance. Talking about topic paths coming from a ROS 1 background is somewhat awkward. See 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -38,152 +58,157 @@ A valid name has the following characteristics:
 
-| `123abc`    | `123`  | `foo bar`  | ` `        | `foo//bar` |
-| `/~`        | `~foo` | `foo~`     | `foo~/bar` | `foo/~bar` |
-| `foo/~/bar` | `foo/` | `foo__bar` |            |            |
```
Hmm, this (`foo__bar`) was in the "not valid name" list, but I didn't see it in the bullet points any where. I know it was part of a previous alternative we tried (described at the bottom of the document), but I wonder if the code checks for this still...

wjwwood replied:
Looks like not, so forget my previous comment: https://github.com/ros2/rcl/blob/21ff57b14b6f406d948f6e9ea97db75a1b2004ef/rcl/src/rcl/validate_topic_name.c#L97-L99

hidmic replied:
I cleaned up the examples table and removed the `foo__bar` entry in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -38,152 +58,157 @@ A valid name has the following characteristics:
+A relative name does not begin with a forward slash (`/`) and it is resolved _relative_ to the addressing node's namespace.
+Periods (`.`) in the addressing node's namespace are replaced by forward slashes (`/`) and a leading forward slash (`/`) is added during resolution.
+
+For example, if a `ping.pong.ball` node addresses (e.g. subscribes to) a `foo/bar` topic, then the topic name will be resolved to `/ping/pong/foo/bar`.
```
```suggestion
For example, if a node with the address `ping.pong.ball` publishes or subscribes to a topic `foo/bar`, then the topic name will be resolved to `/ping/pong/foo/bar`.
```

hidmic replied:
Corrected in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -38,152 +58,157 @@ A valid name has the following characteristics:
-The bracket syntax (`{substitution_name}`) may be used in non-fully qualified names to substitute useful contextual information into the name.
-The set of substitution keys (names) are not set in this document, but some reasonable examples might be: `{node}` expands to the current node's name or `{ns}` expands to the current node's namespace.
+Substitutions must result in a valid name and comply with the name constraints.
+An example of an invalid substitution would be `{sub}/foo` and replace `{sub}` with a numeric value, which thus leads to a (invalid) resource name starting with a numeric character.
```
```suggestion
An example of an invalid substitution would be `{sub}/foo` and replace `{sub}` with a numeric value, which thus leads to an *invalid* resource name starting with a numeric character.
```

hidmic replied:
Corrected in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -192,81 +217,153 @@ Substitutions are also not allowed to be nested, i.e. substitutions may not cont
-### Hidden Topic or Service Names
+#### Fully Qualified Resource Names (FQRNs)
+
+Regular resource names resolution rules allow for convenient syntax, which in some cases requires additional context to expand to a fully qualified name that can be mapped to e.g. a DDS topic name.
```
```suggestion
Regular resource name resolution rules allow for convenient syntax, which in some cases requires additional context to expand to a fully qualified name that can be mapped to another type of name, e.g. a DDS topic name.
```

hidmic replied:
Corrected in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -192,81 +217,153 @@ Substitutions are also not allowed to be nested, i.e. substitutions may not cont
+#### Hidden nodes
 
-## Mapping of ROS 2 Topic and Service Names to DDS Concepts
+If a (resource provider) node name contains any tokens that start with an underscore (`_`), the node it refers to will be considered hidden and tools may not show them unless explicitly asked to.
```
```suggestion
If a node name (a.k.a. resource provider) contains any tokens that start with an underscore (`_`), the node it refers to will be considered hidden and tools may not show them unless explicitly asked to do so.
```

hidmic replied:
Corrected in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -192,81 +217,153 @@ Substitutions are also not allowed to be nested, i.e. substitutions may not cont
-Previously forward slashes (`/`) were disallowed in DDS topic names, now the restriction has been lifted (see [issue](https://issues.omg.org/issues/lists/dds-rtf5#issue-42236) on omg.org) and therefore the ROS topic names are first prefixed with ROS Specific Namespace prefix (described below) are then mapped directly into DDS topic names.
+#### Hidden resources
+
+If a resource name contains any tokens that start with an underscore (`_`), the resource it refers to will be considered hidden and tools may not show them unless explicitly asked to.
```
```suggestion
If a resource name contains any tokens that start with an underscore (`_`), the resource it refers to will be considered hidden and tools may not show them unless explicitly asked to do so.
```

hidmic replied:
Corrected in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -192,81 +217,153 @@ Substitutions are also not allowed to be nested, i.e. substitutions may not cont
+- must not start with a numeric character (`[0-9]`).
+- must not end nor start with a period (`.`).
+- must not contain any number of repeated periods (`.`).
+- must not contain any number of repeated underscores (`_`).
```
So this is back :D. We should figure out what we're doing here and be consistent. Unless I missed it this wasn't a constraint mentioned above. I think we may have wanted to keep it to allow us to have a "safe" delimiter to use in the future, but it looks like the code doesn't check this, at least for topics.

hidmic replied:
Made this constraint explicit everywhere else in 7680627ee8f2fa8675fd58e72f814585c6721f2a. Do you think we shouldn't?

wjwwood replied:
Hmm, not sure. I think we only _needed_ it for the alternative we're no longer using. I don't think the code enforces it at the moment. Maybe it should be removed. @sloretz or anyone on the @ros2/team do you have an opinion here?

jacobperron replied:
I don't have a strong preference. But it seems okay to remove the constraint if there's no reason to have it. If we really need it in the future, we could start using repeated `.` as a delimiter since it's already disallowed.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -192,81 +217,153 @@ Substitutions are also not allowed to be nested, i.e. substitutions may not cont
+
+#### Allowed characters
+
+ROS URL syntax allows for more characters than e.g. DDS topic names allow: the forward slash (`/`), the tilde (`~`), the balanced curly braces (`{}`), the period (`.`) and the semicolon (`:`).
```
```suggestion
The ROS URL syntax allows for more characters than are allowed in DDS topic names.
For example, the forward slash (`/`), the tilde (`~`), the balanced curly braces (`{}`), the period (`.`) and the semicolon (`:`) are all allowed in the ROS URL syntax, but are not allowed in DDS topic names.
```

and actually we are using `/` in the DDS topic names, we just have a understanding with the DDS vendors that in the future they will update the standard to allow at least `/` in addition to the current constraints. And for now, all the implementations allow us to use them.

wjwwood replied:
I see your not below now, maybe we should mention that earlier, like in the DDS topic name restrictions overview near the beginning of the article.

hidmic replied:
Corrected phrasing, moved the note and took away forward slashes as a forbidden character in DDS topic names in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

wjwwood commented on articles/140_resource_names.md:
```diff
@@ -192,81 +217,153 @@ Substitutions are also not allowed to be nested, i.e. substitutions may not cont
+#### Allowed characters
+
+ROS URL syntax allows for more characters than e.g. DDS topic names allow: the forward slash (`/`), the tilde (`~`), the balanced curly braces (`{}`), the period (`.`) and the semicolon (`:`).
+These must be substituted or otherwise removed during the process of converting a ROS URL to e.g. a DDS topic name.
```
```suggestion
These must be substituted or otherwise removed during the process of converting a ROS URL to another kind of name, e.g. a DDS topic name.
```

hidmic replied:
Corrected in 7680627ee8f2fa8675fd58e72f814585c6721f2a.

hidmic committed:
Address peer review comments

- Enable redirects to avoid breaking existing permalinks
- Improve overall writing and internal consistency
- Added additional citations

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

jacobperron commented on articles/140_resource_names.md:
```diff
@@ -1,10 +1,11 @@
 abstract:
-  This article describes the proposed mapping between ROS topic and service names to DDS topic and service names.
-author: '[William Woodall](https://github.com/wjwwood)'
+  This article describes an specification proposal for resource addressing in ROS, as well as its implementation in a DDS setting.
```
```suggestion
  This article describes a specification proposal for resource addressing in ROS, as well as its implementation in a DDS setting.
```

hidmic replied:
Corrected in c753fe9dd2f9bdfe03f92245cab4c80b932a70ed.

jacobperron commented on articles/140_resource_names.md:
```diff
@@ -192,81 +219,155 @@ Substitutions are also not allowed to be nested, i.e. substitutions may not cont
+- must not contain any number of repeated forward slashes (`/`).
+- must not contain any number of repeated underscores (`_`).
+- must separate a tilde (`~`) from the rest of the name with a forward slash (`/`), i.e. `~/foo` not `~foo`.
+- must have balanced curly braces (`{}`) when used, i.e. `{sub}/foo` but not `{sub/foo` nor `/foo}`.
```
Is this point redundant with L280?

hidmic replied:
Good catch. Removed in c753fe9dd2f9bdfe03f92245cab4c80b932a70ed.

jacobperron commented on articles/140_resource_names.md:
```diff
@@ -192,81 +219,155 @@ Substitutions are also not allowed to be nested, i.e. substitutions may not cont
+That is, for example the length of a service name has tighter limits than the length of a topic.
+In the RTI Connext implementation, service names are suffixed with the GUID value of the DDS participant for the service response topic.
+Additionally, a content filtered topic (max length 256 characters) is created which is mapped from the suffixed service name.
+Therefore when linking against `rmw_connext_c` or `rmw_connext_cpp`, service names cannot be longer than 185 characters including the namespace hierarchy and any ros specific prefixes.
```
```suggestion
Therefore when linking against `rmw_connext_c` or `rmw_connext_cpp`, service names cannot be longer than 185 characters including the namespace hierarchy and any ROS specific prefixes.
```

hidmic replied:
Corrected in c753fe9dd2f9bdfe03f92245cab4c80b932a70ed.

hidmic committed:
Address more peer review comments.

Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

mjcarroll commented:
@hidmic What is the status on this?  Can it be merged or does it need to move forward to the Foxy board?

hidmic commented:
It has yet to be implemented, so if anything it has to go to the Foxy board. It was brought up during our brainstorming for Foxy, but AFAIK it has a rather low priority.

hidmic commented:
https://github.com/ros2/ros2/issues/881 to track progress on this topic.
